### ISSUE TITLE
"renders implicit tileset with multiple contents" fails for `ModelExperimental`

### ISSUE THREAD

**Issue by j9liu** [CesiumGS/cesium#10551]
In `Cesium3DTileset`, the "renders implicit tileset with multiple contents" test and its legacy counterpart fail when using `ModelExperimental`. Tis is because the tileset is considered loaded / ready even though all of the contents have not finished loading -- that is, some of the models have `ready = false` when the unit test executes.

**Comment by ptrgags** [comment]
#10540 disabled all the tests that use those same tilesets while we look into it.

**Comment by j9liu** [comment]
#10557 -- same issue with tilesets using implicit multiple contents, with metadata

**Comment by ptrgags** [comment]
Update: @j9liu, @lilleyse and I have been looking into this a bit this morning and at least found a few leads:

* The failure seems easier to reproduce in Firefox than in Chrome.
* when the test fails, at least one of the `Multiple3DTileContents` is unable to immediately schedule requests due to throttling by `RequestScheduler` (limit is currently 6, and there are 9 contents in the tileset). 
* in the unit tests, `waitForTilesLoaded()` waits until `tileset.tilesLoaded` is true. This happens at least once between when the implicit tileset transcodes the tiles into `Cesium3DTile` objects and when the tileset load promise for the test resolves. The tileset uses the statistics object to determine when the tiles are loaded; it doesn't directly wait on any ready promise.
* When the tileset load promise in the test resolves, not all of the contents are ready. This might have to do with the timing of transcoding implicit tiles) and waiting for those tiles' contents to be ready.
* @lilleyse and I have a hunch that it has to do with the removal of `defer()` promises in #10361. I'm currently checking if I can reproduce the problem before vs after that branch was merged. I'll post the results of this testing soon.

The TL;DR is the tileset is incorrectly detecting when the tiles are loaded, and it might have to do with how `Multiple3DTileContent` manages the promises when requests are throttled.

**Comment by ptrgags** [comment]
Results from testing the `defer()` changes. I enabled ModelExperimental for `Cesium3DTileset` specs and put a breakpoint where the failure would happen. (all of this in Firefox)

Commit before the merge:
- I ran the tests 10 times (given that this error is sometimes hard to reproduce). I was not able to reproduce the error
Commit after the merge:
- On the first try, I was able to reproduce the error

So that seems to confirm that the removal of `defer()` was problematic. 

I think it might have to do with these lines: https://github.com/CesiumGS/cesium/pull/10361/files#diff-c017c6fa6370bc4f50ba103f37694a9dc12498d25dc26f8834e30b4cdfca2df5R192-R199 -- If `readyPromise` is accessed too soon, the contents list might be empty so `Promise.all()` will resolve immediately even though the contents haven't even been fetched.

I'm going to try rewriting the promise in an `initialize()` function and ensure it only resolves when the inner contents both exist and are ready.

**Comment by ptrgags** [comment]
Update: still struggling with this one. I see that my idea of using `initialize()` isn't necessarily a good idea (it would essentially be the same defer anti-pattern).

I do notice that there's a race condition between `tileset.allTilesLoaded` and when the tile state is set to READY, often the `allTilesLoaded` event triggers first, regardless of Model vs ModelExperimental, defer or no defer. Not sure if this is expected behavior; it seems odd to me.

![image](https://user-images.githubusercontent.com/8422414/179308754-a9658d5f-b392-409a-8c17-225f8735821e.png).

[Sandcastle for testing](http://localhost:8080/Apps/Sandcastle/index.html#c=zVh7Uxs3EP8qqicTzhNzBhImrW2YNoY2tFCY2JP8gZlEvltjFVm6SjqDw/i7d6V76Rwn4ZWZehhOJ+3jtw+tVtcHzdJZ+A7+TUGbQTSFOOWgwhm9YbN0ls/rM1ADUHNQZI+87I7ESERSaEPmDK7dpIBr0s9kvXdzwagRufe+FIYyAWrUaJHbkSBET2ksr3WHGJVCy85ccjmGDplQrmEklk3UkEkO4caAiINcdDaZvbw8GDIO+kjoBCIj1Qm7YcJyZshYMW/hnMgYOKLMhUZrBTjpjtIZWNByGV2FUaoUCDNkM6gb+2fKGRUH1ECw82r39e7OTri9++r1LztNz0s6AgGVeveKyxwMMRYCmBWXFnidu3IS9Ne5fSXZrP0JOkOvjRrDjAIdXKwo0DJVEXSKCYJk7QEaqtsIlrZrPmwfzRLOImbKwUnKDUs42OCh3bqdg/i4HW6H/2gpSl3L1tdAKSo0T9HW7wILwzb+DSgqhzXo3lBj87J4eoILWPeH1D5NKKb2D0OWicecvCdE5/jNSHKpnhZb34rU90STb3/yXvJ09lhn5VlaDj4wM81qRa4m03JPiLl5T+ssC80Nh3TM74sIC4qhIno0plJONXoaXO1TxTBFqWEe25Nh9ITfE2BfzhKpmXlsopVyvNH9kJxJJkyfy/SxMawEecN3f7x5MJ6+dH42/SeoEGvB1RQ8GObfUs3wMP8RAHPRD4b2NAVjLbSiijwU2oGikfwRwJzgb8G6cP81cOyE8BzLCDskFTFMsHeLW+v6tqXrWvJG6EpgnyRTExrUdRWUXYzfBxkp+ZjadjGWEZZ6YcJLMIcc7PDN4ijGpjGnGTUs46psmiR88YaJmIlLXeloFZKb6xBZ6FbP6VhjC2sLp885auRmoUZHqdOxjhQbQzBJRWTLGAlAWLa4WQSvaOQyTrQnJ+ja5eXDUNR9/y00MrFPXaJhE1I0yHm4gjzUzWaVbq7rDBPFZsywOehQwUzOoSTt5vlQyfxpRWih1xP6ZZcd5vLQKWX2dKuENqkSlaZsUHH4DbIPlcaYGsW2qZrvWuaDCW6rfZIq3iE54LDYR61qPQvYAYzTyw9MwUS5zVheSHwiZ9bhTQKIx55sPKMryJbNbJQ5sGZRSDl32I4ljSG2ZhzOUcQx09hV4x0paJK9/cqbdpdIDiGXl8EG8mZyCHfcG2WEsvzy9Sig8eJMyRnTJa7QTEF4WeNF7Ttxq+4kBXkel89SzoYyqOZLBC1/zovQWwSGe/WMYWF8R8Ul1JgJ2WrV3zd3wtWpE2qm9joabG9thVtkszR6LFNXCAbJFBSECjWlukWQqOlL8F4qxxXeNiRREiNrGOjK8rCa9Jxgt4QveGV3VDxN8vz5nQjRPexyanyEfpCq+Gqz4CsXTz/3B3Y5uK37zd0hOmRl1pkdM7cvygul/zsfNZ7dZriWZH+P/PzS3tvxMm/FBRtJqvDc2bBu3m2OGhetO0jYqklQNpXvxvm6zimVzaC7Mm/v1JgXwLm8vrPmGi/HjX9ntTXOaEHFNzhdHfHpxzz9iqbVmWXtfdn0EnW5WpqwGES2NfFPNKWkqmWbmSp5Tdx8txLgn2jleRWuHFZ55a7VEX2+deEd/VMqYr76pWgQKQAxSGgEri6+zYiC4tMLFXOqM905P2o2RyJJzW/ODs8ge5bZ8pzblGmdAMUDx+6cNQUvX3QWrjlC8+XqtKvVC2xWbMHIiezBflYtBEU0MhYO4tJMkdrjDbPJnG4iFQns5yCGVFtdfPRyri558YI160eED6Iu9ZxddNcdJp+e3VZUyw55drsGeFCRNJefav0A/lu2vha14SKB8Pjw9+HH/vFR/6//TbwKE/UU83ov+7jYvZs1J0cHB8eHlT2NVqPnavB+4d1fGd4slbF9RoDtuAFsx6nB/nucRldYsSOtCw/22j5rL2ZzwuK9NZ9HScSp1rgySTkfsM9YGvZ7baT/gtV2BHjync5BcbqwZNPt/eNsMgzDXhtf13OW7XVpSS/bzCTGu8TmGHtrpMobp0757bFV9FJDuMF7wYa9u+AhMKdYrzqr94aSuE/dE+n7Uyk1EEpyEgS54azLWCswFnOP2YwhBiNh3TSF6Gosb9A/PkQ3DXGHlO37Phlkw5rl3vg/)

I do see that requests get canceled sometimes when the requests are throttled (as expected, there are more than 6 contents). In the sandcastle, it does seem to gracefully handle this case, though it may take a few frames to schedule the request.

**Comment by ptrgags** [comment]
Trying it more, I can reproduce the problem just by running a single failing test, but it doesn't happen every single time

(and even when running all three failing tests together, sometimes only one or two of them fail).

So clearly the throttling is causing some sort of race condition. And from my testing the other day, it seems like the removal of `defer()` is relevant (either it caused the race condition or made it more likely to happen).

I'll keep poking around through the promises to see what's happening.

**Comment by ptrgags** [comment]
Update: logging select rendering statistics to the console, I'm definitely noticing that sometimes the 3 variables that affect `tilesLoaded` (numberOfPendingRequests, numberOfTilesProcessing, numberOfAttemptedRequest) hit 0 before all the contents are ready.

At least one case I saw the number of pending requests increase after the `tilesLoaded` event happened, even though the number of attempted requests stayed at 0 for most of the run. Perhaps throttled/canceled requests isn't being correctly detected in some corner case?

When `defer()` was removed, some promise code is triggered multiple times. the short-circuiting logic gets confusing so still wrapping my head around it.

I also noticed that `numberOfTilesTotal` was incorrect for multiple contents. It's a 1-line fix, though it only impacts the 3D Tiles Inspector.
